hub/core/compression.py

```diff
@@ -805,311 +760,137 @@ def _read_audio_shape(
-    file: Union[bytes, memoryview, str], compression: Optional[str]
-) -> dict:
-    duration = None
-    command = [
-        ffprobe_binary(),
-        "-select_streams",
-        "v:0",
-        "-show_entries",
-        "stream=width,height,duration,avg_frame_rate",
-        "-of",
-        "default=noprint_wrappers=1",
-        "pipe:",
-    ]
+    seek_target = _frame_to_stamp(start, vstream)
+    step_time = _frame_to_stamp(step, vstream)
 
-    if isinstance(file, str):
-        command[-1] = file
-        pipe = sp.Popen(command, stdout=sp.PIPE, stderr=sp.PIPE, bufsize=10**5)
-        raw_info, raw_err = pipe.communicate()
-        duration = bytes.decode(re.search(DURATION_RE, raw_err).groups()[0])  # type: ignore
-        duration = to_seconds(duration)
-    else:
-        if file[: len(_HUB_MKV_HEADER)] == _HUB_MKV_HEADER:
-            mv = memoryview(file)
-            n = len(_HUB_MKV_HEADER) + 2
-            duration = struct.unpack("f", mv[n : n + 4])[0]
-            file = mv[n + 4 :]
-        pipe = sp.Popen(
-            command, stdin=sp.PIPE, stdout=sp.PIPE, stderr=sp.PIPE, bufsize=10**5
-        )
-        raw_info = pipe.communicate(input=file)[0]
-    ret = dict(
-        map(lambda kv: (bytes.decode(kv[0]), kv[1]), re.findall(INFO_RE, raw_info))
-    )
-    ret["width"] = int(ret["width"])
-    ret["height"] = int(ret["height"])
-    if compression in (
-        "mp4",
-        "mkv",
-    ):  # use file values instead of stream values for consistency in case of mp4, mkv
-        ret["duration"] = duration
-    elif "duration" in ret:
-        ret["duration"] = float(ret["duration"])
+    gop_size = (
+        vstream.codec_context.gop_size
+    )  # gop size is distance (in frames) between 2 I-frames
+    if step > gop_size:
+        step_seeking = True
     else:
-        ret["duration"] = duration
-    rate_fraction = map(float, ret["rate"].split(b"/"))
-    ret["rate"] = next(rate_fraction) / next(rate_fraction)
-    return ret
-
-
-def to_seconds(time):
-    return sum([60**i * float(j) for (i, j) in enumerate(time.split(":")[::-1])])
-
-
-def to_hub_mkv(file: str):
-    command = [
-        ffmpeg_binary(),
-        "-i",
-        file,
-        "-codec",
-        "copy",
-        "-f",
-        "matroska",
-        "pipe:",
-    ]
-    pipe = sp.Popen(
-        command, stdin=sp.PIPE, stdout=sp.PIPE, stderr=sp.PIPE, bufsize=10**5
-    )
-    mkv, raw_info = pipe.communicate()
-    duration = bytes.decode(re.search(DURATION_RE, raw_info).groups()[0])  # type: ignore
-    duration = to_seconds(duration)
-    mkv = _HUB_MKV_HEADER + struct.pack("<Hf", 4, duration) + mkv
-    return mkv
+        step_seeking = False
 
+    seekable = True
+    try:
+        container.seek(seek_target, stream=vstream)
+    except av.error.PermissionError:
+        seekable = False
+        container, vstream = _open_video(file)  # try again but this time don't seek
+        print("Cannot seek. Possibly a corrupted video file.")
```

Comment:
use warning instead of print, and mention that we're trying to open full video now